python-package/xgboost/spark/data.py

```diff
@@ -1,192 +1,166 @@
+            base_margin=self._fetch(self.train_data.get(alias.margin, None)),
         )
-        return feature_matrix, y, w, b_m, feature_matrix_val, y_val, w_val, b_m_val
-    return feature_matrix, y, w, b_m
+        self._iter += 1
+        return 1
+
+    def reset(self) -> None:
+        self._iter = 0
 
 
-def _process_data_iter(
-    data_iterator: Iterator[pd.DataFrame],
-    train: bool,
-    has_weight: bool,
+def create_qdm_from_partitions(
+    iterator: Iterator[pd.DataFrame],
     has_validation: bool,
-    has_fit_base_margin: bool = False,
-    has_predict_base_margin: bool = False,
-):
-    """
-    If input is for train and has_validation=True, it will split the train data into train dataset
-    and validation dataset, and return (train_X, train_y, train_w, train_b_m <-
-    train base margin, val_X, val_y, val_w, val_b_m <- validation base margin)
-    otherwise return (X, y, w, b_m <- base margin)
-    """
-    return _row_tuple_list_to_feature_matrix_y_w(
-        data_iterator,
-        train,
-        has_weight,
-        has_fit_base_margin,
-        has_predict_base_margin,
-        has_validation,
-    )
-
-
-def _convert_partition_data_to_dmatrix(
-    partition_data_iter,
-    has_weight,
-    has_validation,
-    has_base_margin,
-    dmatrix_kwargs=None,
-):
-    # pylint: disable=too-many-locals, unbalanced-tuple-unpacking
-    dmatrix_kwargs = dmatrix_kwargs or {}
-    # if we are not using external storage, we use the standard method of parsing data.
-    train_val_data = _prepare_train_val_data(
-        partition_data_iter, has_weight, has_validation, has_base_margin
-    )
-    if has_validation:
-        (
-            train_x,
-            train_y,
-            train_w,
-            train_b_m,
-            val_x,
-            val_y,
-            val_w,
-            val_b_m,
-        ) = train_val_data
-        training_dmatrix = DMatrix(
-            data=train_x,
-            label=train_y,
-            weight=train_w,
-            base_margin=train_b_m,
-            **dmatrix_kwargs,
-        )
-        val_dmatrix = DMatrix(
-            data=val_x,
-            label=val_y,
-            weight=val_w,
-            base_margin=val_b_m,
-            **dmatrix_kwargs,
-        )
-        return training_dmatrix, val_dmatrix
-
-    train_x, train_y, train_w, train_b_m = train_val_data
-    training_dmatrix = DMatrix(
-        data=train_x,
-        label=train_y,
-        weight=train_w,
-        base_margin=train_b_m,
-        **dmatrix_kwargs,
-    )
-    return training_dmatrix
+    features_cols: Sequence[str] = [],
+) -> Tuple[DMatrix, Optional[DMatrix]]:
+    if not features_cols:
+        # features cols must not be empty
+        raise ValueError()
+    # fixme: always on device when this is used.
+    it = PartIter(iterator, features_cols, True, has_validation)
+    dvalid = it.dvalid
+    return DeviceQuantileDMatrix(it), dvalid
```

Comment:
Is `DeviceQuantileDMatrix` only used in GPU training ?

Reply:
Shall we replace CPU side DMatrix with https://github.com/dmlc/xgboost/issues/7890 ?

Reply:
Does Quantile dmatrix reduce training result precision ?

Reply:
https://github.com/dmlc/xgboost/issues/7890 is still working in progress and it's only applicable to hist tree method. I will update it once ready.

> Does Quantile dmatrix reduce training result precision

For CPU no. But for GPU it might generate slight different result due to the fact that GPU uses sliding window for quantile summary generation while the CPU processes a single sample at a time.